keps/sig-scheduling/5004-dra-extended-resource/README.md

```diff
@@ -0,0 +1,835 @@
+
+```go
+type Resource struct {
+	MilliCPU         int64
+	Memory           int64
+	EphemeralStorage int64
+	// We store allowedPodNumber (which is Node.Status.Allocatable.Pods().Value())
+	// explicitly as int, to avoid conversions and improve performance.
+	AllowedPodNumber int
+	// ScalarResources
+	ScalarResources map[v1.ResourceName]int64
+
+    // NEW!
+    // DynamicResources: keep track of dynamic extended resources
+	DynamicResources map[v1.ResourceName]int64
+}
+```
+
+type `NodeInfo` is used by scheduler to keep track of the information for each
+node in memory. Its `Allocatable` field is used to keep track of the allocatable
+resources in memory. For a node with extended resources, its NodeInfo's
+Allocatable.ScalarResources is updated with the `Node`'s informer, minus the
+used. For a node with dynamic extended resources, its NodeInfo's
+Allocatable.DynamicResources is updated with the `ResourceSlice`'s informer,
+minus used by either dynamic extended resource or resource claims.
+
+The existing 'noderesources' plugin needs to be modified, such that a pod's
+extended resource request is checked against a NodeInfo's ScalarResources if the
+node uses device plugin, and checked against a NodeInfo's DynamicResources if
+the node uses DRA driver and it advertised the extended resource in its
+`ResourceSlice`.
+
+The existing 'dynamicresources' plugin needs no modification. Since the special
+resource claim for dynamic extended resource has accounted for the resource
+usage.
+
+A new scheduler plugin 'dynamic extended resource' is added. It uses the
+SharedDRAManager, which is shared with the 'dynamicresources' plugin.
+
+The new plugin implements the following extension points.
+
+#### EventsToRegister
+This registers all cluster events that might make an unschedulable pod schedulable,
+like finishing the allocation of a claim, or resource slice updates.
+
+#### Score
+If two nodes can fit the pod, one node has installed device plugin, the other
+has installed DRA, score the DRA node higher.
+
+#### Reserve
+Allocate the specific devices for the number of dynamic extended resource
+requested in the pod, create an in-memory `ResourceClaim` for storing the
+allocation results. The claim is stored in the map of in-flight claims.
+
+#### Prebind
+This is called in a separate goroutine. The plugin makes API call to create the
+`ResourceClaim` and updates the pod's status. If some API request fails now,
+PreBind fails and the pod must be retried.
+
+#### Unreserve
+The plugin removes the Pod from the claim.status.reservedFor field, removes the
+special `ResourceClaim` for dynamic extended resource. Because it cannot be scheduled
+after all.
+
+### Actuation for Dynamic Extended Resource
+When a pod with dynamic extended resources is picked up by the kubelet on the node it
+is scheduled to run, the following are particularly important:
+
+1. Kubelet tries to admit the pod, the pod's dynamic extended resources requests
```

Comment:
```suggestion
1. Kubelet tries to admit the pod, the pod's dynamic extended resources requests
```

There wouldn't be any of those, just `ScalarResources` and claims.

Reply:
do you mean just avoid introducing 'dynamic extended resources' ? it's ok to avoid that. however, the admission change below is still needed.

Reply:
I think we can drop `Resource.DynamicResources` and in the text stop using the term "dynamic extended resources": "extended resource request" is the existing API that we want to support, it doesn't suddenly change into something else.

Reply:
Which admission change?

Reply:
Oh, the admission check in the kubelet. Yes, that's still needed, depending on what was allocated.

Reply:
Resource.DynamicResources is introduced for scheduler, such that the plugin 'noderesource' can keep (don't filter out) the node with devices in ResourceSlice for the pod with 'extended resource request'. Agreed that on the consumption side, the request is the same as extended resource. However, on the capacity side, it is very different, in that Node.Capacity/Allocatable does not show it at all.